src/main/java/org/elasticsearch/cluster/DiffableUtils.java

```diff
@@ -0,0 +1,204 @@
+     */
+    public static <T extends Diffable<T>> Diff diff(ImmutableMap<String, T> before, ImmutableMap<String, T> after) {
+        assert after != null && before != null;
+        Map<String, Diff> diffs = newHashMap();
+        List<String> deletes = newArrayList();
+        Map<String, T> adds = newHashMap();
+        for (String key : before.keySet()) {
+            if (!after.containsKey(key)) {
+                deletes.add(key);
+            }
+        }
+        for (ImmutableMap.Entry<String, T> partIter : after.entrySet()) {
+            T beforePart = before.get(partIter.getKey());
+            if (!partIter.getValue().equals(beforePart)) {
+                if (beforePart == null) {
+                    adds.put(partIter.getKey(), partIter.getValue());
+                } else {
+                    diffs.put(partIter.getKey(), partIter.getValue().diffs(beforePart));
+                }
+            }
+        }
+        return new CompositeDiff<>(deletes, diffs, adds);
+    }
+
+
+    /**
+     * Reads Diff object produced by {@link org.elasticsearch.cluster.DiffableUtils#diff(org.elasticsearch.common.collect.ImmutableOpenMap, org.elasticsearch.common.collect.ImmutableOpenMap)} method
+     *
+     * The diffs are applied to the before object.
+     */
+    public static <T extends Diffable<T>> ImmutableOpenMap<String, T> readDiff(ImmutableOpenMap<String, T> before, StreamInput in, final T proto) throws IOException {
+        return readDiff(before, in, new KeyedReader<T>() {
+            @Override
+            public T readFrom(StreamInput in, String key) throws IOException {
+                return proto.readFrom(in);
+            }
+        });
+    }
+
+    /**
+     * Reads Diff object produced by {@link org.elasticsearch.cluster.DiffableUtils#diff(org.elasticsearch.common.collect.ImmutableOpenMap, org.elasticsearch.common.collect.ImmutableOpenMap)} method
+     *
+     * The diffs are applied to the before object.
+     */
+    public static <T extends Diffable<T>> ImmutableOpenMap<String, T> readDiff(ImmutableOpenMap<String, T> before, StreamInput in, KeyedReader<T> keyedReader) throws IOException {
+
+        ImmutableOpenMap.Builder<String, T> builder = ImmutableOpenMap.builder();
+        builder.putAll(before);
+
+        int deletesSize = in.readVInt();
+        for (int i = 0; i < deletesSize; i++) {
+            builder.remove(in.readString());
+        }
+
+        int diffsSize = in.readVInt();
+        for (int i = 0; i < diffsSize; i++) {
+            String key = in.readString();
+            builder.put(key, builder.get(key).readDiffs(in));
+        }
+
+        int addsSize = in.readVInt();
+        for (int i = 0; i < addsSize; i++) {
+            String key = in.readString();
+            builder.put(key, keyedReader.readFrom(in, key));
+        }
+        return builder.build();
+    }
+
+    /**
+     * A reader that can deserialize an object
```

Comment:
can we be slightly more verbose on why this is use full? (different deserialization logic based on key in parent map)